### PR TITLE
Raise warning once only if trainable_weights and _collected_trainable_weights are inconsistent

### LINKED ISSUE #22012: [Bug] Discrepancy in tf.keras and keras in setting model.trainable = False and then compiling
### System information
- Have I written custom code: Yes
- OS Platform and Distribution: Linux Ubuntu 16.04
- TensorFlow installed from: binary
- TensorFlow version: v1.10.1-0-g4dcfddc5d1 1.10.1
- Bazel version: N/A
- CUDA/cuDNN version: CUDA9.1, cuDNN7.0
- GPU model and memory: TITAN V
- Exact command to reproduce: python3 compare.py
- Mobile device: N/A

### Describe the problem
This is a toy case for training GAN problem.
When I run the code shown below in keras2.2.2, I get
```
WARNING:tensorflow:Discrepancy between trainable weights and collected trainable weights, did you set `model.trainable` without calling `model.compile` after ?
```
only at the beginning of the training once.
However, I run this code in tensorflow1.10.1, the warning raises at every iteration.
Although it seems that the model is appropriately learned (I can make sure the weight is freezed by the result of `.summary()`), raising too many warning is not torelable.

Here is the complete log.
[tf.keras ver](https://github.com/tensorflow/tensorflow/files/2342979/tf_keras.log)
[keras ver](https://github.com/tensorflow/tensorflow/files/2342980/keras.log)

I see the same problem is posted in Stackoverflow
https://stackoverflow.com/questions/50468940/tensorflow-1-8-tf-keras-gives-different-result-in-dcgan-from-keras

```
import numpy as np

# use keras
from keras.layers import Dense, Input
from keras.models import Model

# use tf.keras
# from tensorflow.keras.layers import Dense, Input
# from tensorflow.keras.models import Model

# define input
noise = Input(shape=(10,))
x = Input(shape=(100,))

# define generator and discriminator
gen = Dense(100)
dis = Dense(1)

y = dis(x)
dis_model = Model(x, y)
dis_model.compile(optimizer='rmsprop', loss='mse')
dis_model.summary()

z = dis_model(gen(noise))
dis_model.trainable = False
combined_model = Model(noise, z)
combined_model.compile(optimizer='rmsprop', loss='mse')
combined_model.summary()

for i in range(3):
    dis_model.train_on_batch(x=np.random.rand(10, 100),
                             y=np.random.rand(10, 1))
    combined_model.train_on_batch(x=np.random.rand(10, 10),
                             y=np.random.rand(10, 1))

```

### PR DESCRIPTION
**Pull request by yanboliang**
Fix #22012
If ```trainable_weights``` and ```_collected_trainable_weights``` are inconsistent, raise warning only once, not every batch(```train_on_batch```) which would flood the screen. ```train_on_batch``` inside loop is not efficient as each batch would rebuild the ```train_function```, but it's the most simple way to train model like GAN in #22012 . This may not a perfect fix, but at least, don't flood my screen. The updated behavior would be the same as ```keras-team/keras```.

### CODE REVIEW

**Review by fchollet — APPROVED** [review]
That makes sense, thanks for the PR.

Btw shouldn't duplicate warnings always be suppressed?

**Comment by yanboliang** [comment]
@fchollet It seems ```tf.logging``` only suppress warning if you explicitly call ```log_first_n```. Thanks.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @facaiy: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
